lua/metals/config.lua

```diff
@@ -37,24 +37,45 @@ local function in_disabled_mode(config)
   return disabled
 end
 
+--- Used to generate Metals installation file name depending on configuration
+--- @return string fileame
+local function metals_filename()
+  if config_cache and config_cache.settings.metals.useGlobalExecutable then
+    return metals_name
+  else
+    return Path:new(util.nvim_metals_cache_dir, metals_name).filename
+  end
+end
+
 --- Ultimately what will be passed to the config.cmd to initialize the LSP
 --- connection. If a user is using useGlobalExecutable then we just default
 --- to `metals`, if not we take control and construct the location in the
 --- cache dir.
 --- @return string executable
 local function metals_bin()
+  local metals_filename = metals_name
+  -- Coursier will install both a "metals" and "metals.bat", on Windows we should use "metals.bat" as exe
+  local metals_win_filename = metals_name .. ".bat"
+
+  if util.has_bins(metals_win_filename) then
+    metals_filename = metals_win_filename
+  end
+
   if config_cache and config_cache.settings.metals.useGlobalExecutable then
-    return metals_name
+    return metals_filename
   else
-    return Path:new(util.nvim_metals_cache_dir, metals_name).filename
+    return Path:new(util.nvim_metals_cache_dir, metals_filename).filename
   end
 end
 
 --- Check to see if coursier is installed. This method favors the native cs. So if
 --- cs is installed, that will be returned, if not, then coursier will be returned.
 --- @return string 'cs', 'coursier', or nil
 local function check_for_coursier()
-  if util.has_bins("cs") then
+  -- this must be first the second "cs" is also found as binary but we should be more specific on Windows
+  if util.has_bins("cs.bat") then
```

Comment:
So there is a few different places that we are checking for `cs.bat` or `metals.bat` even when we aren't on windows. Let's not do this, but instead check the os, and only look for the bat versions if we are on window. You can do this with `vim.loop.os_uname()` which returns a table of info that you can look at. You'll probably want to look at `sysname` in there.

Reply:
a bit complicated given my [non existent] Lua skills but will try.

Reply:
No worries. I do it a different way in the `.jvmopts` file, but I'll extract that into a utility function and change it later.